### PR TITLE
avoid creating objects with duplicate names

### LINKED ISSUE #1839: variables with duplicate names are added to the model anyway
```julia

julia> m = Model();

julia> @variable(m, x)
x

julia> @variable(m, x)
ERROR: An object of name x is already attached to this model. If this is intended, consider using the anonymous construction syntax, e.g., x = @variable(model, [1:N], ...) where the name of the object does not appear inside the macro.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] register_object(::Model, ::Symbol, ::VariableRef) at /home/mlubin/.julia/dev/JuMP/src/JuMP.jl:620
 [3] top-level scope at /home/mlubin/.julia/dev/JuMP/src/macros.jl:322

julia> all_variables(m)
2-element Array{VariableRef,1}:
 x
 x
```

Reported by @mtanneau

### PR DESCRIPTION
**Pull request by mlubin**
Closes #1839

### CODE REVIEW

File: src/macros.jl [hunk 1]
```diff
@@ -303,25 +303,44 @@ function _macro_return(code, variable)
...
+        $(if model_for_registering !== nothing
+            :(_error_if_cannot_register($model_for_registering,
+                                        $(quot(name))))
+          end)
         $variable = $macro_code
         $(if model_for_registering !== nothing
-              :(register_object($model_for_registering, $(quot(name)),
-                $variable))
+            :(object_dictionary($model_for_registering)[$(quot(name))] =
+              $final_variable)
```
Comment (blegat) [inline src/macros.jl]: Should it be `variable` instead of `final_variable` ?

Comment (mlubin) [reply src/macros.jl]: The only case I found where `final_variable` is different from `variable` is here:
https://github.com/JuliaOpt/JuMP.jl/blob/2239b48ab4bb4b753c712a9080f6343545e29fbb/src/macros.jl#L1383

We want the lookup to return the `Symmetric` object in this case, right?

Comment (blegat) [reply src/macros.jl]: Indeed, we should have a test for that

Comment (mlubin) [reply src/macros.jl]: Opened https://github.com/JuliaOpt/JuMP.jl/issues/1842.

**Pull request merged**